Added caching using caffeine

sw-tracker opened the pull request:
Hi Ben,

I have added the caching functionality to the SDK. You'll be able to see that it doesn't add any further dependencies as it is very light weight. I followed the same pattern you guys implemented with the configuration for the HTTP client.

```
final FlagsmithClient flagsmithClient = FlagsmithClient.newBuilder()
                .setApiKey(envKey)
                .enableLogging(FlagsmithLoggerLevel.INFO)
                .withConfiguration(FlagsmithConfig
                        .newBuilder()
                        .baseURI(this.baseURI)
                        .build())
                .withCache(FlagsmithCacheConfig
                        .newBuilder()
                        .maxSize(10)
                        .build())
                .build();
```

I will create a PR in the Flagsmith-docs to document how the cache can be configured once this PR is accepted.

Cheers,
Olga

dabeeeenster commented:
Thanks for this - we will take a look!

matthewelwell commented on src/test/java/com/flagsmith/FlagsmithCachedClientTest.java:
```diff
@@ -0,0 +1,356 @@
+    assignTraitToUserIdentity(user2.getIdentifier(), "unchanged", "stable-value", environment.apiKey);
+
+    assertThat(environment.client.getTrait(user2, traitKey))
+        .isEqualTo(trait(null, traitKey, "xxx"));
```
Perhaps I'm missing something but shouldn't the first argument to the `trait` method here be the identifier of `user2` ?

matthewelwell replied:
(similarly for the rest of the calls to `trait()` which provide `null` as their first argument)

matthewelwell replied:
Sorry to be a pain but I'm not sure why this one got resolved? I can't see any changes in this area.

sw-tracker replied:
`assertThat(environment.client.getTrait(user2, traitKey)).isEqualTo(trait(null, traitKey, "xxx"));`

Do you mean why is it checking in the line above that the trait identity is `null`?
The reason is that, that is what the Flagsmith API returns. This is an integration test so it is validating against what the API returns. If you were expecting the identity to be there, then you will need to change the API (it isn't an SDK issue).

You can see that it has been like this for a long time. Tests (that I didn't write nor changed) in FlagsmithClientTest.java have the same behaviour. Check test:

- testClient_When_Add_Traits_For_Identity_Then_Success lines 448, 449, 454, 455
- testClient_When_Add_Traits_For_Identity_To_Existing_Identity_Then_Success lines 468, 469, 474, 475

matthewelwell reviewed (commented):
Hi @sw-tracker, thanks very much for this, it's really great. 

I've added a few comments in a few places where I've either not quite understood the code or where I think some improvements could be made but I really think it's not far off being ready to merge. 

One general thing that I have a slight issue with is the naming of the relevant classes. To me `Endpoints` indicates a collection of well, endpoints, not API requests (or wrappers around them). I'm not entirely sure of how to improve the naming but one for you to think about please and I'll do the same.

sw-tracker commented on src/main/java/com/flagsmith/FlagsmithCachedEndpoints.java:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  @Override
+  public FlagsmithCache getCache() {
```
I Have added a description. Essentially it returns an object that encapsulates methods to manipulate the cache without exposing the cache object itself.

sw-tracker commented on src/main/java/com/flagsmith/FlagsmithClient.java:
```diff
@@ -557,13 +424,30 @@ public Builder withApiUrl(String apiUrl) {
         public FlagsmithClient build() {
-            client.defaultConfig = this.configuration;
-            return client;
+            FlagsmithEndpoints flagsmithEndpoints = new FlagsmithEndpoints(this.configuration, this.customHeaders, client.logger, apiKey);
```
The instantiated object is used in both the if and else blocks.

sw-tracker commented on src/test/java/com/flagsmith/FlagsmithCachedClientTest.java:
```diff
@@ -0,0 +1,356 @@
+    // Act: cache remains with 2 entries
+    environment.client.getUserFlagsAndTraits(user3);
+    clientCache.cleanUp();
+    assertEquals(2, clientCache.estimatedSize());
```
Done

sw-tracker commented on src/main/java/com/flagsmith/FlagsmithClient.java:
```diff
@@ -557,13 +424,30 @@ public Builder withApiUrl(String apiUrl) {
+
+        /**
+         * Enable in-memory caching for the Flagsmith API.
+         * If no other cache configuration is set, the Caffeine defaults will be used, i.e. no limit
```
I was in two minds about this. In the end I decided to leave the Caffeine defaults and document it in Flagsmith-docs. If you prefer we could add that if the user does not set any of the parameters, then we set one of them. Thoughts?

sw-tracker commented on src/main/java/com/flagsmith/FlagsmithCachedEndpoints.java:
```diff
@@ -0,0 +1,55 @@
+    }
+    // cache exists but does not match the target trait, lets invalidate this cache entry
+    cache.getCache().invalidate(user.getIdentifier());
+    return flagsmithEndpoints.postUserTraits(user, toUpdate, doThrow);
```
Not here, at least not yet. Right now we only cache responses that have FlagsAndTraits. This request only returns a Trait and not the Flags.

matthewelwell replied:
Makes sense, thanks for explaining.

sw-tracker commented on src/test/java/com/flagsmith/FlagsmithCachedClientTest.java:
```diff
@@ -0,0 +1,356 @@
+    final List<Flag> featureFlags1Again = environment.client.getFeatureFlags(user1);
+    assertEquals(1, clientCache.estimatedSize());
+    assertNotNull(featureFlags1Again);
+    assertEquals(featureFlags1, featureFlags1Again);
```
Added unit tests for this.

sw-tracker commented on src/test/java/com/flagsmith/FlagsmithCachedClientTest.java:
```diff
@@ -0,0 +1,356 @@
+    final FlagsAndTraits flagsAndTraits1Again = environment.client.getUserFlagsAndTraits(user1);
+    assertEquals(1, clientCache.estimatedSize());
+    assertNotNull(flagsAndTraits1Again);
+    assertEquals(flagsAndTraits1, flagsAndTraits1Again);
```
Added a remote feature flag modification that should not be in the cache.

matthewelwell replied:
Thanks, that makes sense. So I guess the line `assertEquals(1, clientCache.estimatedSize());` is now checking to confirm that we're using the cache, right? I think it would be good to add a comment / some comments to explain that a little bit.

sw-tracker commented:
Hi @matthewelwell,
I have updated the PR with the review feedback. I pushed the changes in a new commit so you can see what changed (I didn't squash).

Regarding the 2 open points:
- `Endpoints`: this is the naming we use where I work. Essentially if we have Microservice A, B and C. In the codebase of A we have 2 classes called `BEndpoints` and `CEndpoints`. I am happy to rename it, I am just not sure what to. `API`?
- Caffeine sane defaults: I was in two minds about this. In the end I decided to leave the Caffeine defaults and document it in Flagsmith-docs so the developers can decide. If you prefer we could add logic so if the developer does not set any of the parameters, then we set one of them, maybe maximumSize=10. Thoughts?

I have added unit tests to the cached endpoints. There are no unit tests for other classes and currently there are `integration` tests in `FlagsmithClientTest` that should be unit tests (those testing null input parameters). A clean-up task for another day ;)

Cheers,
Olga

matthewelwell commented:
Thanks @sw-tracker and apologies for the delay in getting back to you. I've added a couple of responses to some of the comments I added last time but nothing major. 

 * Regarding the naming convention I guess I would probably go with `APIWrapper` or maybe `APIMethods` which is slightly closer to the `Endpoints` naming convention that you use but just a slight tweak to indicate a group of actions rather than a group of constants.
 * Regarding the defaults, if caffeine has defaults then that's fine but my understanding was that the default was 'no limit' which seems slightly dangerous to me.

That's fine regarding the tests - fully appreciate all the work you've put in to test everything in this PR plus some! We definitely need to tidy up the testing on the other bits - it's on our list :D.

Thanks,
Matt

sw-tracker reviewed (commented):
`assertThat(environment.client.getTrait(user2, traitKey)).isEqualTo(trait(null, traitKey, "xxx"));`

Do you mean why is it checking in the line above that the trait identity is `null`?
The reason is that, that is what the Flagsmith API returns. This is an integration test so it is validating against what the API returns. If you were expecting the identity to be there, then you will need to change the API (it isn't an SDK issue).

You can see that it has been like this for a long time. Tests (that I didn't write nor changed) in FlagsmithClientTest.java have the same behaviour. Check test:

- testClient_When_Add_Traits_For_Identity_Then_Success lines 448, 449, 454, 455
- testClient_When_Add_Traits_For_Identity_To_Existing_Identity_Then_Success lines 468, 469, 474, 475

sw-tracker commented:
Hi @matthewelwell,

I have renamed the `Endpoints` to `APIWrapper`.

Regarding the defaults, I have set them to max 10 identities for 5 minutes. I also added some unit tests for `FlagsmithCacheConfig`.

I may add in a separate PR a banner to see unit tests coverage, then we can see how well we are doing ;)

Cheers,
Olga

matthewelwell commented:
Hi @sw-tracker, 

This looks good to me now. I've just asked one one of my colleagues to give it a second pair of eyes but we should be able to approve and merge very soon. 

Apologies again for the delay. 

Thanks,
Matt

sw-tracker commented:
Hi @matthewelwell,

Fantastic! Thank you! I have updated the Flagsmith documentation to reflect these changes: https://github.com/Flagsmith/flagsmith-docs/pull/9. Please have a look at that PR.

Cheers,
Olga
